File: maximal_square_BF.java
// TC - O(M^2 N^2)
// SC - O(1)
class Solution {
	public int maximalSquare(char[][] matrix) {
		if (matrix == null || matrix.length == 0)
			return 0;
		int m = matrix.length;
		int n = matrix[0].length;
		int max = 0;
		boolean flag = false;
		for (int i = 0; i < m; i++) {
			for (int j = 0; j < matrix[0].length; j++) {
				if (matrix[i][j] == '1') {
					int curr = 1;
					flag = true;
					while (i + curr < m && j + curr < n && flag) {
						// column
						for (int k = i + curr; k >= i; k--) {
							if (matrix[k][j + curr] == '0') {
								flag = false;
								break;
							}
						}
						// row
						for (int k = j + curr; k >= j; k--) {
							if (matrix[i + curr][k] == '0') {
								flag = false;
								break;
							}
						}
						if (flag)
							curr++;
					}
					max = Math.max(max, curr);
				}
			}
		}
		return max * max;
	}
}

File: maximalSquare_bf.py
"""
Time complexity O(MN)^2
Space complexity O(1)
"""


class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if matrix is None or len(matrix)==0:
            return matrix
        m=len(matrix)
        n=len(matrix[0])
        flag=False
        maxarea=0
        for i in range(m):
            for j in range(n):
                if(matrix[i][j]=='1'):
                    flag=True
                    curr=1
                    while(i+curr<m and j+curr <n and flag):
                        for k in range(i+curr,i-1,-1):
                            if(matrix[k][j+curr]=='0'):
                                flag=False
                                break
                        for k in range(j+curr,j-1,-1):
                            if(matrix[i+curr][k]=='0'):
                                flag=False
                                break
                        if flag!=False:
                            curr+=1
                    maxarea =max(maxarea,curr)
        return maxarea*maxarea  
                            
                            
                                
        